dataset --> https://www.kaggle.com/datasets/uciml/red-wine-quality-cortez-et-al-2009

binary classification

In [2]:
import pandas as pd

In [3]:
df = pd.read_csv('winequality-red.csv')

In [4]:
df.shape

(1599, 12)

In [5]:
df.head()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
0,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5
1,7.8,0.88,0.00,2.6,0.098,25.0,67.0,0.9968,3.20,0.68,9.8,5
2,7.8,0.76,0.04,2.3,0.092,15.0,54.0,0.9970,3.26,0.65,9.8,5
3,11.2,0.28,0.56,1.9,0.075,17.0,60.0,0.9980,3.16,0.58,9.8,6
4,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5


In [ ]:
# for binary classification, we can convert the quality column into a binary variable(0/1)
df['quality'] = df['quality'].apply(lambda x: 1 if x>=5.5 else 0)
df['quality'].value_counts()
# 0- bad wine, 1- good wine

quality
1    855
0    744
Name: count, dtype: int64

In [7]:
df

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
0,7.4,0.700,0.00,1.9,0.076,11.0,34.0,0.99780,3.51,0.56,9.4,0
1,7.8,0.880,0.00,2.6,0.098,25.0,67.0,0.99680,3.20,0.68,9.8,0
2,7.8,0.760,0.04,2.3,0.092,15.0,54.0,0.99700,3.26,0.65,9.8,0
3,11.2,0.280,0.56,1.9,0.075,17.0,60.0,0.99800,3.16,0.58,9.8,1
4,7.4,0.700,0.00,1.9,0.076,11.0,34.0,0.99780,3.51,0.56,9.4,0
...,...,...,...,...,...,...,...,...,...,...,...,...
1594,6.2,0.600,0.08,2.0,0.090,32.0,44.0,0.99490,3.45,0.58,10.5,0
1595,5.9,0.550,0.10,2.2,0.062,39.0,51.0,0.99512,3.52,0.76,11.2,1
1596,6.3,0.510,0.13,2.3,0.076,29.0,40.0,0.99574,3.42,0.75,11.0,1
1597,5.9,0.645,0.12,2.0,0.075,32.0,44.0,0.99547,3.57,0.71,10.2,0


In [9]:
X = df.iloc[:,:-1]
y = df.iloc[:,-1]

In [13]:
print(X.shape, y.shape)

(1599, 11) (1599,)


- normalising the data
- brings all features to similar scale. 
- z = x-mean/std


In [14]:
mean = X.mean(axis = 0)
X -= mean
std = X.std(axis = 0)
X /= std

In [16]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [ ]:
model = Sequential()
model.add(Dense(8, activation='relu', input_shape=(X.shape[1],)))
# 8 neurons in this layer. This is the first hidden layer
# input_dim=len(X[0, :])
# Number of input features
# X.shape[0] = 1599 → number of samples
# X.shape[1] = 11 → number of features
# ReLU is used for hidden layers because:
# Faster training
# Avoids vanishing gradient
model.add(Dense(4, activation='relu'))
model.add(Dense(1, activation='sigmoid'))  # output layer

c:\Users\Mrunmayee Potdar\Desktop\dl\tf_env\lib\site-packages\keras\src\layers\core\dense.py:95: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [19]:
model.summary()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 8)              │            96 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 4)              │            36 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 1)              │             5 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 137 (548.00 B)

 Trainable params: 137 (548.00 B)

 Non-trainable params: 0 (0.00 B)

### 1️⃣ Output Shape

* `(None, 8)` → First layer has **8 neurons**
* `(None, 4)` → Second layer has **4 neurons**
* `(None, 1)` → Output layer has **1 neuron**
* `None` = batch size (number of samples, not fixed)



### 2️⃣ Param # (How Parameters Are Calculated)

**Formula for Dense layer:**

[
\text{Parameters} = (inputs \times neurons) + neurons
]

✔ First layer:
[
(11 \times 8) + 8 = 96
]

✔ Second layer:
[
(8 \times 4) + 4 = 36
]

✔ Output layer:
[
(4 \times 1) + 1 = 5
]



### 3️⃣ Total Parameters

[
96 + 36 + 5 = 137
]

All are trainable → so:

* **Total params = 137**
* **Trainable params = 137**
* **Non-trainable params = 0**



### 🔑 Key Concept

Each neuron has:

* One weight per input
* One bias

So:

```
Total params = (inputs × neurons) + biases
Biases = number of neurons
```






In [20]:
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

- Optimizer → How the model learns
- Loss → What the model tries to minimize
- Metrics → How we check performance

In [21]:
model.fit(X, y, epochs=100)

Epoch 1/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.4790 - loss: 0.7246
Epoch 2/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5397 - loss: 0.6956
Epoch 3/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5679 - loss: 0.6716
Epoch 4/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6023 - loss: 0.6510
Epoch 5/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6404 - loss: 0.6333
Epoch 6/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6604 - loss: 0.6171
Epoch 7/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6785 - loss: 0.6058
Epoch 8/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6873 - loss: 0.5972
Epoch 9/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.6917 - loss: 0.5909
Epoch 10/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6986 - loss: 0.5853
Epoch 11/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7061 - loss: 0.5800
Epoch 12/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy:

In [22]:
predictions = model.predict(X)
print(predictions[:5])

50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step
[[0.23818403]
 [0.13966575]
 [0.30504304]
 [0.61528915]
 [0.23818403]]


In [23]:
print(y[:5])

0    0
1    0
2    0
3    1
4    0
Name: quality, dtype: int64


In [24]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

In [25]:
acc = accuracy_score(y, predictions.round())
prec = precision_score(y, predictions.round())
rec = recall_score(y, predictions.round())
f1 = f1_score(y, predictions.round())

In [26]:
print(acc, prec, rec, f1)

0.7792370231394622 0.8251295336787565 0.7450292397660818 0.7830362630608482


Accuracy → Overall correct

Precision → How correct are positive predictions?

Recall → How many positives did we catch?

F1 → Balance of precision & recall